File: app/src/main/java/com/leafchild/scopequotas/data/Quota.java
package com.leafchild.scopequotas.data;

import com.j256.ormlite.dao.ForeignCollection;
import com.j256.ormlite.field.DatabaseField;
import com.j256.ormlite.field.ForeignCollectionField;
import com.j256.ormlite.table.DatabaseTable;
import com.leafchild.scopequotas.common.Utils;

import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

import static android.R.attr.type;
import static com.leafchild.scopequotas.data.QuotaType.DAILY;
import static com.leafchild.scopequotas.data.QuotaType.MONTHLY;
import static com.leafchild.scopequotas.data.QuotaType.WEEKLY;

/**
 * Created by: leafchild
 * Date: 03/04/2017
 * Project: ScopeQuotas
 */

@DatabaseTable(tableName = Quota.TABLE_NAME_QUOTA)
public class Quota {

    final static String TABLE_NAME_QUOTA = "quota";

    @DatabaseField(generatedId = true)
    private long id;
    @DatabaseField(columnName = "name", canBeNull = false, unique = true)
    private String name;
    @DatabaseField(columnName = "quotaType", canBeNull = false)
    private QuotaType quotaType;
    @DatabaseField(columnName = "description")
    private String description;
    @DatabaseField(columnName = "min")
    private Integer min;
    @DatabaseField(columnName = "max")
    private Integer max;
    @DatabaseField(columnName = "createdDate")
    private Date createdDate;
    @DatabaseField(columnName = "modifiedDate")
    private Date modifiedDate;
    @DatabaseField(columnName = "category", foreign = true, foreignAutoRefresh = true)
    private QuotaCategory category;

    @ForeignCollectionField(columnName = "logged", eager = true)
    private ForeignCollection<Worklog> logged;

    public Quota() {
    }

    public Quota(String name, String description, QuotaType type) {
        this.name = name;
        this.description = description;
        this.quotaType = type;
    }

    @Override
    public String toString() {
        //Have to use name only for spinners
        return name;
    }

    public ForeignCollection<Worklog> getLogged() {
        return logged;
    }

    public void setLogged(ForeignCollection<Worklog> logged) {
        this.logged = logged;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public QuotaType getQuotaType() {
        return quotaType;
    }

    public void setQuotaType(QuotaType quotaType) {
        this.quotaType = quotaType;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Date getCreatedDate() {
        return createdDate;
    }

    public void setCreatedDate(Date createdDate) {
        this.createdDate = createdDate;
    }

    public Date getModifiedDate() {
        return modifiedDate;
    }

    public void setModifiedDate(Date modifiedDate) {
        this.modifiedDate = modifiedDate;
    }

    @Override
    public boolean equals(Object o) {
        if(this == o) {
            return true;
        }
        if(o == null || getClass() != o.getClass()) {
            return false;
        }

        Quota quota = (Quota) o;
        return Objects.equals(quota, this);

    }

    @Override
    public int hashCode() {
        return Objects.hash(name, quotaType, id);
    }

    public boolean isNew() {
        return id <= 0;
    }

    public Float getAllWorklogAmount() {

        float sum = 0;

        for(Worklog w : logged) {
            sum += Utils.transformWorklog(w);
        }
        return Float.valueOf(String.format(Locale.getDefault(), "%.2f", sum));
    }

    public Float getWorkFlowByLastPeriod() {
        Calendar from = Calendar.getInstance();
        Calendar to = Calendar.getInstance();

        switch(this.getQuotaType()) {
            case DAILY:
                from.set(Calendar.HOUR_OF_DAY, 0);
                break;
            case WEEKLY:
                from.set(Calendar.HOUR_OF_DAY, 0);
                from.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY);
                break;
            case MONTHLY:
                from.set(Calendar.HOUR_OF_DAY, 0);
                from.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY);
                from.set(Calendar.MONTH, from.get(Calendar.MONTH));
                break;
            default:
                return 0f;
        }

        return Utils.calculateAmount(logged, from.getTime(), to.getTime());
    }

    public void addWorklog(Worklog worklog) {
        logged.add(worklog);
    }

    public QuotaCategory getCategory() {
        return category;
    }

    public void setCategory(QuotaCategory category) {
        this.category = category;
    }

    public Integer getMax() {
        return max;
    }

    public void setMax(Integer max) {
        this.max = max;
    }

    public Integer getMin() {
        return min;
    }

    public void setMin(Integer min) {
        this.min = min;
    }
}

File: app/src/main/java/com/leafchild/scopequotas/data/Quota.kt
package com.leafchild.scopequotas.data

import com.j256.ormlite.dao.ForeignCollection
import com.j256.ormlite.field.DataType
import com.j256.ormlite.field.DatabaseField
import com.j256.ormlite.field.ForeignCollectionField
import com.j256.ormlite.table.DatabaseTable
import com.leafchild.scopequotas.common.Utils
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by: leafchild
 * Date: 03/04/2017
 * Project: ScopeQuotas
 */

@DatabaseTable(tableName = Quota.TABLE_NAME_QUOTA)
class Quota {

    @DatabaseField(generatedId = true)
    var id: Long = 0
    @DatabaseField(columnName = "name", canBeNull = false, unique = true)
    var name: String? = null
    @DatabaseField(columnName = "quotaType", canBeNull = false)
    var quotaType: QuotaType? = null
    @DatabaseField(columnName = "description")
    var description: String? = null
    @DatabaseField(columnName = "min")
    var min: Int? = null
    @DatabaseField(columnName = "max")
    var max: Int? = null
    @DatabaseField(columnName = "createdDate")
    var createdDate: Date? = null
    @DatabaseField(columnName = "modifiedDate")
    var modifiedDate: Date? = null
    @DatabaseField(dataType = DataType.BOOLEAN_OBJ, columnName = "archived", defaultValue = "false")
    var archived: Boolean? = null
    @DatabaseField(columnName = "category", foreign = true, foreignAutoRefresh = true)
    var category: QuotaCategory? = null

    @ForeignCollectionField(columnName = "logged", eager = true)
    var logged: ForeignCollection<Worklog>? = null

    val isNew: Boolean
        get() = id <= 0

    val allWorklogAmount: Float?
        get() {

            var sum = 0f

            for (w in logged!!) {
                sum += Utils.transformWorklog(w)!!.toFloat()
            }
            return java.lang.Float.valueOf(String.format(Locale.getDefault(), "%.2f", sum))
        }

    val workFlowByLastPeriod: Float?
        get() {

            val from = Calendar.getInstance(Locale.UK)
            val to = Calendar.getInstance(Locale.UK)

            when (this.quotaType) {
                QuotaType.DAILY -> from.set(Calendar.HOUR_OF_DAY, 0)
                QuotaType.WEEKLY -> {
                    from.set(Calendar.HOUR_OF_DAY, 0)
                    from.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY)
                    to.set(Calendar.HOUR_OF_DAY, 23)
                    to.set(Calendar.DAY_OF_WEEK, Calendar.SUNDAY)
                }
                QuotaType.MONTHLY -> {
                    from.set(Calendar.HOUR_OF_DAY, 0)
                    from.set(Calendar.DAY_OF_MONTH, 0)
                }
                else -> {
                    from.set(Calendar.HOUR_OF_DAY, 0)
                    from.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY)
                    to.set(Calendar.HOUR_OF_DAY, 23)
                    to.set(Calendar.DAY_OF_WEEK, Calendar.SUNDAY)
                }
            }

            return Utils.calculateAmount(logged!!, from.time, to.time)
        }

    constructor() {}

    constructor(name: String, description: String, type: QuotaType) {

        this.name = name
        this.description = description
        this.quotaType = type
    }

    override fun toString(): String {
        //Have to use name only for spinners
        return name!!
    }

    override fun equals(o: Any?): Boolean {

        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }

        val quota = o as Quota?
        return quota!!.hashCode() == this.hashCode()

    }

    override fun hashCode(): Int {

        return Objects.hash(name, quotaType, id)
    }

    fun toExportString(): Array<String> {

        val forExport = ArrayList<String>()
        forExport.add(id.toString())
        forExport.add(name!!)
        forExport.add(description!!)
        forExport.add(min.toString())
        forExport.add(max.toString())
        forExport.add(SimpleDateFormat("dd-MM-yyyy", Locale.US).format(modifiedDate))
        forExport.add(category!!.name!!)
        forExport.add(allWorklogAmount.toString())
        forExport.add(archived!!.toString().toUpperCase())

        return forExport.toTypedArray()
    }

    companion object {

        internal const val TABLE_NAME_QUOTA = "quota"
    }
}


